src/radical/pilot/agent/agent_0.py

```diff
@@ -160,25 +160,32 @@ def _configure_app_comm(self):
     #
     def initialize(self):
 
+        # we  first register outputs, then start known service tasks, and only
+        # once those are up and running we register inputs.  This ensures that
+        # we do not start any incoming tasks before we services are up.
+
+        # forward received tasks to agent input staging
+        self.register_output(rps.AGENT_STAGING_INPUT_PENDING,
+                             rpc.AGENT_STAGING_INPUT_QUEUE)
+
+        # and return completed tasks to the task manager
+        self.register_output(rps.TMGR_STAGING_OUTPUT_PENDING,
+                             rpc.PROXY_TASK_QUEUE)
+
+        # start any services
+        self._start_services()
+
         # listen for new tasks from the client
         self.register_input(rps.AGENT_STAGING_INPUT_PENDING,
                             rpc.PROXY_TASK_QUEUE,
                             qname=self._pid,
                             cb=self._proxy_input_cb)
 
-        # and forward to agent input staging
-        self.register_output(rps.AGENT_STAGING_INPUT_PENDING,
-                             rpc.AGENT_STAGING_INPUT_QUEUE)
-
         # listen for completed tasks to forward to client
         self.register_input(rps.TMGR_STAGING_OUTPUT_PENDING,
                             rpc.AGENT_COLLECTING_QUEUE,
                             cb=self._proxy_output_cb)
 
-        # and register output
-        self.register_output(rps.TMGR_STAGING_OUTPUT_PENDING,
-                             rpc.PROXY_TASK_QUEUE)
-
         self.register_rpc_handler('prepare_env', self._prepare_env,
                                                  rpc_addr=self._pid)
 
```

Comment:
should these lines L189-L202 be moved before service tasks start? in case of `rp` env is used by service tasks

Reply:
https://github.com/radical-cybertools/radical.pilot/blob/c0ba7f1883cfd1fff2bf5eac2c336752c4d83ab0/src/radical/pilot/agent/agent_0.py#L189-L202

Reply:
that makes sense, will do.

Reply:
The RPC handler itself is actually not useful: that can only handle `pilot.prepare_env` calls *after* the pilot became active.  We also have to pass the `prepare_env` attributes from the pilot description - that is done now.